torch/nn/functional.py

```diff
@@ -4841,28 +4841,79 @@ def _in_projection(
 
 scaled_dot_product_attention = _add_docstr(
     torch._C._nn.scaled_dot_product_attention, r"""
+scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.0, is_causal=False) -> Tensor:
+
 Computes scaled dot product attention on query, key and value tensors, using
 an optional attention mask if passed, and applying dropout if a probability
 greater than 0.0 is specified.
 
+.. code-block:: python
+
+    # Composite Implementation
+    attn_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=0) if is_causal else attn_mask
+    attn_mask = attn_mask.masked_fill(not attn_mask, -float('inf')) if attn_mask.dtype==torch.bool else attn_mask
+    attn_weight = torch.softmax((Q @ K.transpose(-2, -1) / math.sqrt(Q.size(-1))) + attn_mask, dim=-1)
+    attn_weight = torch.dropout(attn_weight, dropout_p)
+    return attn_weight @ V
+
+.. warning:: This function is beta and subject to change.
+
+Note:
+    This function calls into one of three backends:
+        * `FlashAttention: Fast and Memory-Efficient Exact Attention with IO-Awareness`_
+        * `Memory-Efficient Attention`_
+        * A pytorch implementation defined in c++ matching the above mathematical formulation
+
+    The function defaults to selecting the highest-performing implementation based on the inputs provided.
+    However, each of the fused kernels has specific input limitations.
+    If you require a specific backend to be utilized, there exists functions to enable or disable specific backends.
+    Please note that all backends are enabled by default.
+
+    For example :func:`~torch.backends.cuda.enable_flash_sdp` can be used to enable/disable FlashAttention.
+    The context manager :func:`~torch.backends.cuda.sdp_kernel` can be used to enable/disable the backends
+    for a specific scope.
+
+    If a user wants to enforce that one of the fused implementations is used, disable the math fallback
+    using one of the above mechanisms. If for some reason a fused implementation is not available,
+    the function will throw an error with the reasons why the fused implementation was not used.
+
+    The numerical accuracy of the fused kernels has been tested but due to the nature of fusing floating point operations
+    The deviations from the infinite precision implementation may be significant. If that is the case we encourage users
+    to please file an issue. A work around would be disabiling the fused kernels and using the math fallback. For more
```

Comment:
math fallback here again.
Also do you actually expect the math impl to be more precise?

Reply:
we have called it the third option math I.e.: https://github.com/pytorch/pytorch/blob/3a5a762443da7e733b4f6c6cf9574da92507967f/torch/backends/cuda/__init__.py#L229

the math impl supports fp64 so in that case yes. I think its more nature of the fp errors is not exactly consistent between fused kernels and the composite one so if you trained your model using the composite one it might perform slightly differently if you are now using the fused kernels

Reply:
The different versions being different makes sense.
What I was surprised is that the phrasing hints at the fact that the math one is more precise is surprising.
But if you say: the math version with double precision inputs is the most precise. That sounds good!